File: .claude/skills/run-magic/driver.mjs
#!/usr/bin/env node
// CDP browser driver for the Magic LEA web app.
//
// No npm dependencies: uses Node 22's built-in global WebSocket + fetch to
// speak the Chrome DevTools Protocol directly. `chromium-cli`/Playwright are
// NOT installed in this environment, so this is the drive-the-browser harness.
//
// Subcommands (see SKILL.md for the canonical invocations):
//   shot   <url> <outfile>              navigate, settle, PNG screenshot
//   eval   <url> <jsExpr>               navigate, print JSON result of jsExpr
//   click  <url> <selector> <outfile>   navigate, click selector, screenshot
//   flow   <outfile>                    full scripted flow: start a human-vs-AI
//                                        game from the menu, screenshot the board
//
// Env:
//   APP_URL      base app url (default http://127.0.0.1:8010)
//   CHROME       path to chrome/edge exe (auto-detected on Windows if unset)
//   HEADED=1     launch a visible window instead of headless

import { spawn } from 'node:child_process';
import { mkdtempSync, writeFileSync, existsSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

const APP_URL = process.env.APP_URL || 'http://127.0.0.1:8010';
const PORT = 9222 + (process.pid % 1000);

function findChrome() {
  if (process.env.CHROME) return process.env.CHROME;
  const candidates = [
    'C:/Program Files/Google/Chrome/Application/chrome.exe',
    'C:/Program Files (x86)/Google/Chrome/Application/chrome.exe',
    'C:/Program Files (x86)/Microsoft/Edge/Application/msedge.exe',
    'C:/Program Files/Microsoft/Edge/Application/msedge.exe',
  ];
  for (const c of candidates) {
    try { if (existsSync(c)) return c; } catch {}
  }
  return candidates[0];
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function launchChrome() {
  const userDir = mkdtempSync(join(tmpdir(), 'mtg-cdp-'));
  const headless = process.env.HEADED ? [] : ['--headless=new', '--disable-gpu'];
  const args = [
    `--remote-debugging-port=${PORT}`,
    `--user-data-dir=${userDir}`,
    '--no-first-run',
    '--no-default-browser-check',
    '--window-size=1400,900',
    ...headless,
    'about:blank',
  ];
  const proc = spawn(findChrome(), args, { stdio: 'ignore' });
  // Poll the version endpoint until the debug server answers.
  for (let i = 0; i < 100; i++) {
    try {
      const r = await fetch(`http://127.0.0.1:${PORT}/json/version`);
      if (r.ok) { await r.json(); return proc; }
    } catch {}
    await sleep(100);
  }
  throw new Error('Chrome devtools endpoint never came up');
}

// Minimal CDP client over a single page target.
class CDP {
  constructor(ws) { this.ws = ws; this.id = 0; this.pending = new Map(); }
  static async attach() {
    // Find a page target (Chrome opens one for about:blank).
    let target;
    for (let i = 0; i < 50; i++) {
      const list = await (await fetch(`http://127.0.0.1:${PORT}/json/list`)).json();
      target = list.find((t) => t.type === 'page');
      if (target) break;
      await sleep(100);
    }
    if (!target) throw new Error('no page target');
    const ws = new WebSocket(target.webSocketDebuggerUrl);
    await new Promise((res, rej) => { ws.onopen = res; ws.onerror = rej; });
    const cdp = new CDP(ws);
    ws.onmessage = (ev) => {
      const msg = JSON.parse(ev.data);
      if (msg.id && cdp.pending.has(msg.id)) {
        const { res, rej } = cdp.pending.get(msg.id);
        cdp.pending.delete(msg.id);
        msg.error ? rej(new Error(JSON.stringify(msg.error))) : res(msg.result);
      }
    };
    return cdp;
  }
  send(method, params = {}) {
    const id = ++this.id;
    this.ws.send(JSON.stringify({ id, method, params }));
    return new Promise((res, rej) => this.pending.set(id, { res, rej }));
  }
  async navigate(url) {
    await this.send('Page.enable');
    await this.send('Runtime.enable');
    await this.send('Page.navigate', { url });
    // Wait for the app's JS to settle. The app boots its menu synchronously,
    // so a fixed settle plus a DOM-ready poll is reliable enough.
    await sleep(800);
    for (let i = 0; i < 30; i++) {
      const r = await this.eval('document.readyState');
      if (r === 'complete') break;
      await sleep(150);
    }
    await sleep(400);
  }
  async eval(expr) {
    const r = await this.send('Runtime.evaluate', {
      expression: `(()=>{ try { return JSON.stringify(${expr}); } catch(e){ return JSON.stringify({__err:String(e)}); } })()`,
      returnByValue: true, awaitPromise: true,
    });
    const v = r.result?.value;
    try { return JSON.parse(v); } catch { return v; }
  }
  async click(selector) {
    return this.eval(`(()=>{const el=document.querySelector(${JSON.stringify(selector)}); if(!el) return {clicked:false}; el.click(); return {clicked:true};})()`);
  }
  async shot(outfile) {
    const r = await this.send('Page.captureScreenshot', { format: 'png' });
    writeFileSync(outfile, Buffer.from(r.data, 'base64'));
    return outfile;
  }
}

async function main() {
  const [cmd, ...rest] = process.argv.slice(2);
  const chrome = await launchChrome();
  let exitCode = 0;
  try {
    const cdp = await CDP.attach();
    if (cmd === 'shot') {
      const [url, out] = rest;
      await cdp.navigate(url || APP_URL);
      console.log('saved', await cdp.shot(out || 'shots/app.png'));
    } else if (cmd === 'eval') {
      const [url, expr] = rest;
      await cdp.navigate(url || APP_URL);
      console.log(JSON.stringify(await cdp.eval(expr), null, 2));
    } else if (cmd === 'click') {
      const [url, sel, out] = rest;
      await cdp.navigate(url || APP_URL);
      console.log('click', await cdp.click(sel));
      await sleep(600);
      console.log('saved', await cdp.shot(out || 'shots/click.png'));
    } else if (cmd === 'flow') {
      const [out] = rest;
      await cdp.navigate(APP_URL);
      console.log('home shown:', await cdp.eval('!!document.querySelector("#homeHostBtn")'));
      // Home -> Host Game page. Selectors are stable element ids from index.html.
      await cdp.click('#homeHostBtn');
      await sleep(500);
      // Mode select defaults to "human_vs_ai"; set it explicitly to be safe,
      // then click "Create Session" (#startBtn).
      console.log('host page:', await cdp.eval('!document.querySelector("#hostGamePage").classList.contains("hidden")'));
      await cdp.eval(`(()=>{const s=document.querySelector('#mode'); s.value='human_vs_ai'; s.dispatchEvent(new Event('change',{bubbles:true})); return s.value;})()`);
      await cdp.click('#startBtn');
      // Session creation hits the API and deals opening hands; give it time.
      await sleep(3500);
      const board = await cdp.eval('!document.querySelector("#boardPanel").classList.contains("hidden")');
      console.log('board visible:', board);
      // The battlefield is canvas-rendered (battlefield-canvas.js), so cards are
      // NOT DOM nodes. Prove the game is live via the prompt/turn text instead.
      console.log('prompt:', await cdp.eval('(document.querySelector("#promptTitle")||{}).textContent || null'));
      console.log('canvas:', await cdp.eval('!!document.querySelector("#battlefieldCanvasWrap canvas")'));
      console.log('saved', await cdp.shot(out || 'shots/flow.png'));
      if (!board) exitCode = 1;
    } else {
      console.error('usage: driver.mjs <shot|eval|click|flow> ...');
      exitCode = 2;
    }
  } catch (e) {
    console.error('ERROR', e.message);
    exitCode = 1;
  } finally {
    try { chrome.kill(); } catch {}
  }
  process.exit(exitCode);
}

main();

File: .claude/skills/run-magic/driver.py
#!/usr/bin/env python
"""Playwright browser driver for the Magic LEA web app.

Drives the canvas-rendered board through a real Chromium that Playwright
manages, so no system Chrome/Edge path detection is needed. Run it with the
workspace venv python (it has `playwright` installed):

    .\\.venv\\Scripts\\python.exe .claude/skills/run-magic/driver.py <cmd> ...

Subcommands (see SKILL.md for the canonical invocations):
  shot   <url> <outfile>              navigate, settle, PNG screenshot
  eval   <url> <jsExpr>               navigate, print JSON result of jsExpr
  evalshot <url> <jsExpr> <outfile>   navigate, run jsExpr, then screenshot
  click  <url> <selector> <outfile>   navigate, click selector, screenshot
  flow   <outfile>                    full scripted flow: start a human-vs-AI
                                       game from the menu, screenshot the board

Env:
  APP_URL    base app url (default http://127.0.0.1:8010)
  HEADED=1   launch a visible window instead of headless
"""
import json
import os
import sys
import time

from playwright.sync_api import sync_playwright

APP_URL = os.environ.get("APP_URL", "http://127.0.0.1:8010")
HEADED = bool(os.environ.get("HEADED"))


def navigate(page, url):
    page.goto(url or APP_URL, wait_until="load")
    # The app boots its menu synchronously; a short settle plus a readyState
    # poll is reliable enough for screenshots/eval.
    page.wait_for_function("() => document.readyState === 'complete'", timeout=10000)
    page.wait_for_timeout(400)


def js_eval(page, expr):
    """Evaluate a page-side JS expression and return its (JSON) value.

    Mirrors the old CDP driver: wraps the expression so a thrown error comes
    back as {"__err": ...} instead of crashing the call. Playwright auto-awaits
    a returned promise.
    """
    wrapped = (
        "(async () => { try { return (%s); } catch (e) { return { __err: String(e) }; } })()"
        % expr
    )
    return page.evaluate(wrapped)


def main():
    args = sys.argv[1:]
    if not args:
        print("usage: driver.py <shot|eval|evalshot|click|flow> ...", file=sys.stderr)
        return 2
    cmd, rest = args[0], args[1:]
    exit_code = 0

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=not HEADED, args=["--window-size=1400,900"])
        page = browser.new_page(viewport={"width": 1400, "height": 900})
        try:
            if cmd == "shot":
                url = rest[0] if rest else APP_URL
                out = rest[1] if len(rest) > 1 else "shots/app.png"
                navigate(page, url)
                page.screenshot(path=out)
                print("saved", out)
            elif cmd == "eval":
                url = rest[0] if rest else APP_URL
                expr = rest[1] if len(rest) > 1 else "document.title"
                navigate(page, url)
                print(json.dumps(js_eval(page, expr), indent=2))
            elif cmd == "evalshot":
                url, expr = rest[0], rest[1]
                out = rest[2] if len(rest) > 2 else "shots/evalshot.png"
                navigate(page, url)
                print("eval", json.dumps(js_eval(page, expr)))
                page.wait_for_timeout(500)
                page.screenshot(path=out)
                print("saved", out)
            elif cmd == "click":
                url, sel = rest[0], rest[1]
                out = rest[2] if len(rest) > 2 else "shots/click.png"
                navigate(page, url)
                clicked = js_eval(
                    page,
                    "(() => { const el = document.querySelector(%s);"
                    " if (!el) return { clicked: false }; el.click();"
                    " return { clicked: true }; })()" % json.dumps(sel),
                )
                print("click", json.dumps(clicked))
                page.wait_for_timeout(600)
                page.screenshot(path=out)
                print("saved", out)
            elif cmd == "flow":
                out = rest[0] if rest else "shots/flow.png"
                navigate(page, APP_URL)
                print("home shown:", js_eval(page, '!!document.querySelector("#homeHostBtn")'))
                # Home -> Host Game page. Selectors are stable element ids from index.html.
                page.click("#homeHostBtn")
                page.wait_for_timeout(500)
                print(
                    "host page:",
                    js_eval(page, '!document.querySelector("#hostGamePage").classList.contains("hidden")'),
                )
                # Mode select defaults to "human_vs_ai"; set it explicitly to be safe,
                # then click "Create Session" (#startBtn).
                js_eval(
                    page,
                    "(() => { const s = document.querySelector('#mode');"
                    " s.value = 'human_vs_ai';"
                    " s.dispatchEvent(new Event('change', { bubbles: true }));"
                    " return s.value; })()",
                )
                page.click("#startBtn")
                # Session creation hits the API and deals opening hands; give it time.
                page.wait_for_timeout(3500)
                board = js_eval(
                    page, '!document.querySelector("#boardPanel").classList.contains("hidden")'
                )
                print("board visible:", board)
                # The battlefield is canvas-rendered (battlefield-canvas.js), so cards are
                # NOT DOM nodes. Prove the game is live via the prompt/canvas instead.
                print(
                    "prompt:",
                    js_eval(page, '(document.querySelector("#promptTitle")||{}).textContent || null'),
                )
                print("canvas:", js_eval(page, '!!document.querySelector("#battlefieldCanvasWrap canvas")'))
                page.screenshot(path=out)
                print("saved", out)
                if not board:
                    exit_code = 1
            else:
                print("usage: driver.py <shot|eval|evalshot|click|flow> ...", file=sys.stderr)
                exit_code = 2
        except Exception as e:  # noqa: BLE001 - report and exit non-zero
            print("ERROR", e, file=sys.stderr)
            exit_code = 1
        finally:
            browser.close()
    return exit_code


if __name__ == "__main__":
    sys.exit(main())
